cpp/src/jit/row_ir.cpp

```diff
@@ -0,0 +1,573 @@
+
+void operation::instantiate(instance_context& ctx, instance_info const& info)
+{
+  for (auto& arg : operands_) {
+    arg->instantiate(ctx, info);
+  }
+
+  id_ = ctx.make_tmp_id();
+  std::vector<data_type> operand_types;
+
+  for (auto& arg : operands_) {
+    operand_types.emplace_back(arg->get_type().type);
+  }
+
+  auto type_id = ast::detail::ast_operator_return_type(op_, operand_types).id();
+
+  // all decimal operation result types should have scale equal to the minimum scale of the operands
+  auto scale = std::accumulate(operand_types.begin() + 1,
+                               operand_types.end(),
+                               operand_types.front().scale(),
+                               [](auto const& a, auto const& b) { return std::min(a, b.scale()); });
+
+  type_ = type_info{cudf::data_type{type_id, scale}};
+}
+
+std::string operation::generate_code(instance_context& ctx,
+                                     target_info const& info,
+                                     instance_info const& instance)
+{
+  std::string operands_code;
+
+  for (auto& arg : operands_) {
+    operands_code =
+      std::format("{}{}{}", operands_code, arg->generate_code(ctx, info, instance), "\n");
+  }
+
+  auto operation_code = [&]() {
+    switch (info.id) {
+      case target::CUDA: {
+        auto first_operand = operands_[0]->get_id();
+        auto operands_str  = (operands_.size() == 1)
+                               ? std::string{first_operand}
+                               : std::accumulate(operands_.begin() + 1,
+                                                operands_.end(),
+                                                std::string{first_operand},
+                                                [](auto const& a, auto& node) {
+                                                  return std::format("{}, {}", a, node->get_id());
+                                                });
+
+        auto cuda = std::format(
+          "{} {} = cudf::ast::detail::operator_functor<cudf::ast::ast_operator::{}, "
+          "false>{{}}({});",
+          cuda_type(type_),
+          id_,
+          ast::detail::ast_operator_string(op_),
+          operands_str);
+        return cuda;
+      }
+      default:
+        CUDF_FAIL("Unsupported target: " + std::to_string(static_cast<int>(info.id)),
+                  std::invalid_argument);
+    }
+  }();
+
+  return operands_code + operation_code;
+}
+
+std::span<ast_input_spec const> ast_converter::get_input_specs() const { return input_specs_; }
+
+int32_t ast_converter::add_ast_input(ast_input_spec in)
+{
+  auto id = static_cast<int32_t>(input_specs_.size());
+  input_specs_.push_back(std::move(in));
+  return id;
+}
+
+std::unique_ptr<row_ir::node> ast_converter::add_ir_node(ast::literal const& expr)
+{
+  auto index = add_ast_input(ast_scalar_input_spec{
+    expr.get_scalar(), expr.get_value(), make_column_from_scalar(expr.get_scalar(), 1)});
```

Comment:
The `make_column_from_scalar` function takes a stream and memory resource parameter.
Should we not honor the stream parameter at least?